notebooks/proteochemometrics.py

```diff
@@ -0,0 +1,300 @@
+            ConvertToNumpyArray(bitvect, arr)
+
+            chembl_id_to_idx[chembl_id] = next(count)
+            tensors.append(torch.tensor(arr))
+            rrr.append(chembl_id)
+
+    tensor = torch.stack(tensors)
+    embedding = Embedding.from_pretrained(tensor)
+
+    return EmbeddingBackmap(embedding, rrr, chembl_id_to_idx)
+
+
+def get_protein_go_triples():
+    """Get triples with proteins as the subject and Gene Ontology terms as the objects."""
+    path = pystow.ensure("bio", "go", url=GO_URL)
+    df = pd.read_csv(
+        path, sep="\t", comment="!", usecols=[0, 1, 3, 4], names=["DB", "DB Object ID", "Relation", "GO ID"]
+    )
+    df = df[~df["Relation"].str.startswith("NOT")]
+    df = df[df["DB"] == "UniProtKB"]
+    df = df[["DB Object ID", "Relation", "GO ID"]].rename(
+        columns={"DB Object ID": "subject", "Relation": "predicate", "GO ID": "object"}
+    )
+    return df
+
+
+def get_chemical_protein_triples():
+    """Get triples from ExCAPE-DB with chemicals as subjects and proteins as objects."""
+    path = pystow.ensure("bio", "excape", url=EXCAPE_URL)
+    ff_path = pystow.join("bio", "excape", name="excape_human_chembl_subset.tsv")
+    if ff_path.is_file():
+        df = pd.read_csv(ff_path, sep="\t")
+        return df
+
+    df = pd.read_csv(
+        path,
+        sep="\t",
+        compression="xz",
+        usecols=["DB", "Original_Entry_ID", "Entrez_ID", "Activity_Flag", "Tax_ID"],
+        dtype=str,
+    )
+
+    # keep only active relationships
+    df = df[df["Activity_Flag"] == "A"]
+    # slice down to human targets
+    df = df[df["Tax_ID"] == "9606"]
+    # keep only relationships qualified by ChEMBL, even though this way out of date in 2025
+    df = df[df["DB"] == "chembl20"]
+
+    df["uniprot"] = df["Entrez_ID"].map(uniprot_client.get_id_from_entrez)
+
+    # throw away anything that can't be mapped to human uniprot
+    df = df[df["uniprot"].notna()]
+
+    df = df[["uniprot", "Original_Entry_ID"]].drop_duplicates().rename(columns={"Original_Entry_ID": "chembl"})
+    df.to_csv(ff_path, sep="\t", index=False)
+    return df
+
+
+def main() -> None:
+    """Demonstrate using chemical representations for a subset of entities."""
+    target_dim = 32
+
+    example_chembl_id = "CHEMBL1097808"
+
+    click.echo("Getting chemical-protein triples from ExCAPE-DB")
+    excape_df = get_chemical_protein_triples()
+    chembl_ids = excape_df["chembl"].unique()
+    uniprot_ids = excape_df["uniprot"].unique()
+
+    if example_chembl_id not in chembl_ids:
+        raise ValueError("need to pick a chembl ID for prediction that's in ExCAPE-DB")
+
+    click.echo("Getting protein representations from UniProt")
+    # example uniprots ["Q13506", "Q13507", "Q13508", "Q13509"]
+    protein_base_repr, _, uniprot_ids = get_human_protein_embedding(uniprot_ids)
+    protein_trans_repr = MLPTransformedEmbedding(protein_base_repr, target_dim)
```

Comment:
From https://github.com/pykeen/pykeen/issues/1505 it sounded like you want to additionally also use a [`CombinedRepresentation`](https://pykeen.readthedocs.io/en/latest/api/pykeen.nn.representation.CombinedRepresentation.html) with an `Embedding`?

Reply:
Maybe it also makes sense to change the order sligthly from

```mermaid
graph LR
   f(Feature)
   e(Embedding)
   c(Combination) 
   t(Learnable Transformation)
   f --> t --> c
   e --> c
``` 

to

```mermaid
graph LR
   f(Feature)
   e(Embedding)
   c(Combination) 
   t(Learnable Transformation)
   f --> c --> t
   e --> c
``` 

so the learnable transformation can also mix the features with the embeddings (while keeping the combination simple, e.g., Concat)?

Reply:
yeah, but on second thought, it seems like a modeling decision that I might want to try a few permutations on. The question is how much/what stuff to upstream into the package itself, versus things that can be ad-hoc in a given example

Reply:
the other question is what happens when some chemicals/proteins have features, both others don't

Reply:
I added another representation module that implements your suggestion in d1daf64c

Reply:
thanks for the suggestion!

I also made the learnable embedding optional in the example because there are millions of chemicals, and it significantly increases training time